test/e2e/framework/helper/certificates.go

```diff
@@ -20,83 +20,151 @@ import (
+			fmt.Fprintf(os.Stderr, "Perceived certificate conditions:\n")
+			for _, cond := range certificate.Status.Conditions {
+				fmt.Fprintf(os.Stderr, "- Last Status: '%s' Reason: '%s', Message: '%s'\n", cond.Status, cond.Reason, cond.Message)
+			}
+		}
+
+		fmt.Fprintf(os.Stderr, "Certificate description:\n")
+		h.Kubectl(certificate.Namespace).DescribeResource("certificate", name)
+		fmt.Fprintf(os.Stderr, "Order and challenge descriptions:\n")
+		h.Kubectl(certificate.Namespace).Describe("order", "challenge")
+
+		fmt.Fprintf(os.Stderr, "Certificaterequest description:\n")
+		crName, err := apiutil.ComputeName(certificate.Name, certificate.Spec)
+		if err != nil {
+			log.Logf("Failed to compute CertificateRequest name from certificate: %s", err)
+		} else {
+			h.Kubectl(certificate.Namespace).DescribeResource("certificaterequest", crName)
+		}
+	}
+	return certificate, pollErr
 }
 
 // WaitForCertificateReady waits for the certificate resource to enter a Ready state and to leave the Issuing state.
 func (h *Helper) WaitForCertificateReady(ns, name string, timeout time.Duration) (*cmapi.Certificate, error) {
-	result, err := e2eutil.WaitForCertificateCondition(h.CMClient.CertmanagerV1().Certificates(ns), name, cmapi.CertificateCondition{
+	ready_true_condition := cmapi.CertificateCondition{
 		Type:   cmapi.CertificateConditionReady,
 		Status: cmmeta.ConditionTrue,
-	}, timeout)
-	if err != nil {
-		return h.handleResult(ns, name, result, "Ready", err)
 	}
-	// Making sure that the Certificate is stable (see #4239) by also waiting for the Issuing state to disappear.
-	// A certificate that has state Ready=True and Issuing not set, is stable and will not change without outside changes.
-	return h.waitForCertificateNotIssuing(ns, name, timeout)
+	issuing_condition := cmapi.CertificateCondition{
+		Type: cmapi.CertificateConditionIssuing,
+	}
+
+	return h.waitPollImmediateCertificate(h.CMClient.CertmanagerV1().Certificates(ns), name, func(certificate *v1.Certificate) bool {
+		if !apiutil.CertificateHasCondition(certificate, ready_true_condition) {
+			log.Logf("Expected Certificate %v condition %v=%v but it has: %v", certificate.Name, ready_true_condition.Type, ready_true_condition.Status, certificate.Status.Conditions)
+			return false
+		}
+
+		if apiutil.CertificateHasCondition(certificate, issuing_condition) {
+			log.Logf("Expected Certificate %v condition %v to be missing but it has: %v", certificate.Name, issuing_condition.Type, certificate.Status.Conditions)
+			return false
+		}
+
+		return true
+	}, 500*time.Millisecond, timeout)
 }
 
 // WaitForCertificateReadyUpdate waits for the certificate resource to enter a
 // Ready state and to leave the Issuing state. If the provided cert was in a
 // Ready state already, the function waits for a state transition to have happened.
 func (h *Helper) WaitForCertificateReadyUpdate(cert *cmapi.Certificate, timeout time.Duration) (*cmapi.Certificate, error) {
```

Comment:
I need some further explanation in the comments for this function and the other `WaitForCertificateReady...` functions explaining the scenarios when I should use the `...Update` variant.

I'm also confused by this line in the documentation:

> If the provided cert was in a Ready state already, the function waits for a state transition to have happened.

It implies to me that the function will always wait for a new Ready=True with an incremented observedGeneration, but I don't think that's the case.

Would it be accurate to call this function `WaitForCertManagerToFinishReconcilingThisCertificate` ? That's how I understand it.  E.g.

Test1 
1. Create certificate
2. WaitForCertManagerToFinishReconcilingThisCertificate
3. Get and Update certificate, safe in the knowledge that we won't encounter an optimistic locking error
4. WaitForCertManagerToFinishReconcilingThisCertificate
5. Poll until cainjector has inject the updated tls.crt into the injectable target resource (safe in the knowledge that cert-manager has updated the Secret contents.

And if that is the use-case, then I don't understand the need for the two variants: `...Ready` and `...ReadyUpdate`

Reply:
> If the provided cert was in a Ready state already, the function waits for a state transition to have happened.

This is the reason the function exists, but it does not correctly explain what the function does or how to use it. I'll update the docs.

Reply:
> And if that is the use-case, then I don't understand the need for the two variants: ...Ready and ...ReadyUpdate

In all cases, ReadyUpdate can be used instead of Ready. I'll maybe try to refactor that.
Note however that the only difference is that this function additionally filters the conditions based on the certificate generation. So, the certificate might have a condition Ready=True but for an older generation and that will not result in the function returning early.